subsys/dfu/src/dfu_target_modem.c

```diff
@@ -0,0 +1,215 @@
+
+	err = setsockopt(fd, SOL_DFU, SO_DFU_APPLY, NULL, 0);
+	if (err < 0) {
+		if (errno == ENOEXEC) {
+			LOG_ERR("SO_DFU_APPLY failed, modem error %d",
+				get_modem_error());
+		} else {
+			LOG_ERR("SO_DFU_APPLY failed, modem error %d", err);
+		}
+	}
+	return 0;
+}
+
+static int delete_banked_modem_fw(void)
+{
+	int err;
+	socklen_t len = sizeof(offset);
+
+	LOG_INF("Deleting firmware image, this can take several minutes");
+	err = setsockopt(fd, SOL_DFU, SO_DFU_BACKUP_DELETE, NULL, 0);
+	if (err < 0) {
+		LOG_ERR("Failed to delete backup, errno %d", errno);
+		return -EFAULT;
+	}
+
+	while (true) {
+		err = getsockopt(fd, SOL_DFU, SO_DFU_OFFSET, &offset, &len);
+		if (err < 0) {
+			if (errno == ENOEXEC) {
+				err = get_modem_error();
+				if (err != DFU_ERASE_PENDING) {
+					LOG_ERR("DFU error: %d", err);
+				}
+				k_sleep(K_MSEC(500));
+			}
+		} else {
+			LOG_INF("Modem FW delete complete");
+			break;
+		}
+	}
+
+	return 0;
+}
+
+/**@brief Initialize DFU socket. */
+static int modem_dfu_socket_init(void)
+{
+	int err;
+	socklen_t len;
+	u8_t version[36];
+
+	/* Create a socket for firmware update */
+	fd = socket(AF_LOCAL, SOCK_STREAM, NPROTO_DFU);
+	if (fd < 0) {
+		LOG_ERR("Failed to open Modem DFU socket.");
+		return fd;
+	}
+
+	LOG_INF("Modem DFU Socket created");
+
+	len = sizeof(version);
+	err = getsockopt(fd, SOL_DFU, SO_DFU_FW_VERSION, &version,
+			    &len);
+	if (err < 0) {
+		LOG_ERR("Firmware version request failed, errno %d", errno);
+		return -1;
+	}
+
+	version[sizeof(version) - 1] = '\0';
+
+	LOG_INF("Modem firmware version: %s", log_strdup(version));
+
+	return err;
+}
+
+bool dfu_target_modem_identify(const void *const buf)
+{
+	return ((struct modem_delta_header *)buf)->magic == MODEM_MAGIC;
+
+}
+
+int dfu_target_modem_init(void)
+{
+	int err;
+	socklen_t len = sizeof(offset);
+
+	err = modem_dfu_socket_init();
+	if (err < 0) {
+		return err;
+	}
+
+	/* Check offset, store to local variable */
+	err = getsockopt(fd, SOL_DFU, SO_DFU_OFFSET, &offset, &len);
+	if (err < 0) {
+		if (errno == ENOEXEC) {
+			LOG_ERR("Modem error: %d", get_modem_error());
+		} else {
+			LOG_ERR("getsockopt(OFFSET) errno: %d", errno);
+		}
+	}
+
+	if (offset == DIRTY_IMAGE) {
+		delete_banked_modem_fw();
+	} else if (offset != 0) {
+		LOG_INF("Setting offset to 0x%x", offset);
+		len = sizeof(offset);
+		err = setsockopt(fd, SOL_DFU, SO_DFU_OFFSET, &offset, len);
+		if (err != 0) {
+			LOG_INF("Error while setting offset: %d", offset);
+		}
+	}
+
+	return 0;
+}
+
+int dfu_target_modem_offset(void)
+{
+	return offset;
+}
+
+int dfu_target_modem_write(const void *const buf, size_t len)
+{
+	int err = 0;
+	int sent = 0;
+	int modem_error = 0;
+
+	sent = send(fd, buf, len, 0);
+	if (sent > 0) {
+		return 0;
+	}
+
+	if (errno == ENOEXEC) {
+		modem_error = get_modem_error();
+		LOG_ERR("send failed, modem errno %d, dfu err %d", errno,
+			modem_error);
+
+		switch (modem_error)
+		{
+			case DFU_INVALID_UUID:
```

Comment:
Align `case` with `switch`. You can simplify by checking `if errno != NOEXEC` so that you can return immediately and save even one more indentation level.